### PR TITLE
add network partition integration test for MaterializeMySQL

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add some test for MaterializeMySQL. e.g. network partition, MySQL kill sync thread...

### CODE REVIEW

File: tests/integration/test_materialize_mysql_database/test.py [hunk 1]
```diff
@@ -20,27 +20,34 @@ def started_cluster():
         cluster.start()
         yield cluster
     finally:
-        cluster.shutdown()
+        #cluster.shutdown()
```
Comment (tavplubix) [inline tests/integration/test_materialize_mysql_database/test.py:23]: Why not to shutdown it?

File: tests/integration/test_materialize_mysql_database/materialize_with_ddl.py [hunk 2]
```diff
@@ -461,3 +463,81 @@ def select_without_columns(clickhouse_node, mysql_node, service_name):
...
+    check_query(clickhouse_node, "SELECT * FROM test_database.test_table FORMAT TSV", '1\n')
+
+    result = mysql_node.query_and_get_data("select id from information_schema.processlist where STATE='Master has sent all binlog to slave; waiting for more updates'")
+
+    for row in result:
+        row_result = {}
+        query = "kill " + str(row[0]) + ";"
+        mysql_node.query(query)
+
+    time.sleep(5)
```
Comment (tavplubix) [inline tests/integration/test_materialize_mysql_database/materialize_with_ddl.py:528]: Tests with `sleep(...)` are usually flacky, but let's look at the result of flacky check.

**Comment by TCeason** [comment]
Oh I will add other case and I forget to modify it.



---Original---

File: tests/integration/test_materialize_mysql_database/materialize_with_ddl.py (hunk 2 above)
Comment (TCeason) [reply tests/integration/test_materialize_mysql_database/materialize_with_ddl.py:528]: If I want to have a MySQL client writing data continuously and then close MySQL Server Docker during the writing process.  Next, stop the client and restart the MySQL Server Docker. Finally check the MySQL data rows and ClickHouse data rows.

Can this finish in integration testing? @tavplubix

**Comment by TCeason** [comment]
In the test case it reports at clickhouse docker shutdown:
```
           Exception: Sanitizer assert found in /ClickHouse/tests/integration/test_materialize_mysql_database/_instances/docker.log for instance [36mnode1_1
```

But I don't know why it occurs. @tavplubix  Is there some good idea to fix it ?

**Comment by tavplubix** [comment]
> In the test case it reports at clickhouse docker shutdown:
> 
> ```
>            Exception: Sanitizer assert found in /ClickHouse/tests/integration/test_materialize_mysql_database/_instances/docker.log for instance �[36mnode1_1
> ```
> 
> But I don't know why it occurs. @tavplubix Is there some good idea to fix it ?

You can find `test_materialize_mysql_database/_instances/docker.log` and other logs in `test_dir.tar`, it contains address sanitizer report: https://gist.github.com/tavplubix/2b536d81d23fc7412a2cf186efda236f

If I understand correctly, it reports stack-use-after-scope of `log_name` [here](https://github.com/ClickHouse/ClickHouse/blob/c5b56c24f83bde0a70e4304bf13d4e723218db67/src/Databases/MySQL/MaterializeMetadata.cpp#L117). Stack-use-after-scope of local variable in the same thread looks very strange, however, asan provides the following hint:
```
 HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
       (longjmp and C++ exceptions *are* supported)
```
It also provides the following sequence of thread creations: 
```
Thread T58 (MySQLDBSync) created by T42 (QueryPipelineEx) here:
    ...
    #11 0x1de7d50f in DB::PipelineExecutor::executeImpl
    ...
Thread T42 (QueryPipelineEx) created by T9 (SystemLogFlush) here:
    ...
    #8 0x1d70c458 in DB::MergeTreeData::loadDataParts
    ...
Thread T9 (SystemLogFlush) created by T0 here:
    ...
    #6 0x1ca4da14 in DB::SystemLog<DB::MetricLogElement>::startup()
    ...
```
It cannot be true, because there is no ways for metric log to create `MaterializeMySQLSyncThread`. 

Maybe we can just add it to asan suppressions list, but I think we should investigate it.

**Comment by TCeason** [comment]
If close weekend the CI will be closed?  @tavplubix

**Comment by tavplubix** [comment]
We have a limitation of maximum number of failed checks. If there are many failed checks, then no new checks are started until all failed tests are fixed. `Marker check` shows the following status in this case: `3 checks failed (max 3)`. `force tests` label means "ignore this limitation and run all tests". 
Recently we have increased this number from 3 to 5 checks to make the limitation more tolerant to flacky tests when failures are probably not related to changes in PR, so now it should work fine without `force tests` label.

**Comment by TCeason** [comment]
> We have a limitation of maximum number of failed checks. If there are many failed checks, then no new checks are started until all failed tests are fixed. `Marker check` shows the following status in this case: `3 checks failed (max 3)`. `force tests` label means "ignore this limitation and run all tests".
> Recently we have increased this number from 3 to 5 checks to make the limitation more tolerant to flacky tests when failures are probably not related to changes in PR, so now it should work fine without `force tests` label.

OK, maybe I should increase my accuracy😂😂

**Comment by TCeason** [comment]
```
=================================== FAILURES ===================================
__________________________ test_network_partition_5_7 __________________________

started_cluster = <helpers.cluster.ClickHouseCluster object at 0x7f9bdab4de80>
started_mysql_5_7 = <test_materialize_mysql_database.test.MySQLNodeInstance object at 0x7f9bd2875ef0>

    def test_network_partition_5_7(started_cluster, started_mysql_5_7):
>       materialize_with_ddl.network_partition_test(clickhouse_node, started_mysql_5_7, "mysql1")

test_materialize_mysql_database/test.py:202: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_materialize_mysql_database/materialize_with_ddl.py:605: in network_partition_test
    check_query(clickhouse_node, "SELECT * FROM test_database.test_table FORMAT TSV", '1\n')
test_materialize_mysql_database/materialize_with_ddl.py:30: in check_query
    assert clickhouse_node.query(query) == result_set
helpers/cluster.py:915: in query
    database=database, ignore_error=ignore_error)
helpers/client.py:21: in query
    password=password, database=database, ignore_error=ignore_error).get_answer()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

```


I see this error is 

```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 210, stderr: Code: 210. DB::NetException: Connection refused (172.18.0.2:9000)

```

But I have already run select 1 on clickhouse to confirm server's network is ok. Is there some other setting about netwokr?


a> add_rule clickhouse_node's 3306 port
b> delete_rule clickhouse_node's 3306 port
c> test for select 1
d> detach db materialize mysql engine
e> attach db
g> select * from db.tab_1  --> err is 9000 port refused.
@tavplubix

**Comment by tavplubix** [comment]
The failure is not related to network settings.
clickhouse-server.log from test_dir.tar: https://gist.github.com/tavplubix/f9a42339429a0c2c042c3991b04336a4

**Comment by TCeason** [comment]
@tavplubix  Integration tests (asan) — Timeout :( Should I re-push a commit to rerun the test?

**Comment by TCeason** [comment]
> @tavplubix Integration tests (asan) — Timeout :( Should I re-push a commit to rerun the test?

@robot-clickhouse
Integration tests (asan) — Timeout :(
Details
@robot-clickhouse
Integration tests (thread) — Timeout :(

timeout again..

**Comment by tavplubix** [comment]
Integration tests failure looks related:
```
        subprocess_call(['docker-compose', 'ps', '--services', '--all'])
>       raise Exception("Cannot wait MySQL container")
E       Exception: Cannot wait MySQL container

helpers/cluster.py:448: Exception
_ ERROR at setup of test_create_and_select_mysql[clickhouse1-complex_node1_cache-LAYOUT(COMPLEX_KEY_CACHE(SIZE_IN_CELLS 10))] _

    @pytest.fixture(scope="module")
    def started_cluster():
        try:
>           cluster.start()
```

**Comment by TCeason** [comment]
https://github.com/ClickHouse/ClickHouse/pull/16806/files#diff-1ff70ba9b166a819d1881dcaa3081cd59ffe7064448b6c3e9263697608d96e1eR5


because of this？

> Integration tests failure looks related:
> 
> ```
>         subprocess_call(['docker-compose', 'ps', '--services', '--all'])
> >       raise Exception("Cannot wait MySQL container")
> E       Exception: Cannot wait MySQL container
> 
> helpers/cluster.py:448: Exception
> _ ERROR at setup of test_create_and_select_mysql[clickhouse1-complex_node1_cache-LAYOUT(COMPLEX_KEY_CACHE(SIZE_IN_CELLS 10))] _
> 
>     @pytest.fixture(scope="module")
>     def started_cluster():
>         try:
> >           cluster.start()
> ```

**Comment by chang0391** [comment]
Is this a solution to resolve the bug of inconsistent data after MySQL or ClickHouse restart?

**Comment by TCeason** [comment]
> Is this a solution to resolve the bug of inconsistent data after MySQL or ClickHouse restart?

Yes, we have already in work.

**Comment by TCeason** [comment]
> Is this a solution to resolve the bug of inconsistent data after MySQL or ClickHouse restart?

This just a test case. Not a solution.

**Comment by chang0391** [comment]
Are you going to resolve the bug?

**Comment by TCeason** [comment]
> Are you going to resolve the bug?

I‘m not sure. It's just in our plan, but not assigned to someone.

**Comment by TCeason** [comment]
It’s ok now.@tavplubix

**Review by tavplubix — APPROVED** [review]

**Pull request merged**